compiler/src/iree/compiler/Codegen/LLVMCPU/DispatchABI.cpp

```diff
@@ -0,0 +1,1082 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/LLVMCPU/DispatchABI.h"
+
+#include "llvm/BinaryFormat/Dwarf.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Path.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+//------------------------------------------------------------------------------
+// ExecutableLibraryDI
+//------------------------------------------------------------------------------
+
+// NOTE: the debug information used here is only present as a compiler developer
+// aid. It may get out of sync with published versions of the executable ABI and
+// may not be very clean. For example, we size push constant and binding arrays
+// not based on the actual layout of the pipeline layout but with some
+// reasonable limit that allows for use in a debugger. If we wanted to improve
+// this we could customize per-scope the structures and look up the
+// IREE::HAL::PipelineLayoutAttr for each entry point to discover the real
+// limits.
+//
+// NOTE: MLIR and subsequent LLVM optimizations will often remove a lot of this
+// debug information (or at least make it less useful). This can happen even in
+// LLVM modes of -O0 as MLIR has no such configurability at this time.
+//
+// It'd be nice to have an automatic sync of the debug information and structs
+// in this file such that we'd get matching source file/line information with
+// the runtime headers and be able to delete most of this hand-authored code.
+// The current debug information and types were constructed by compiling
+// executable_library_demo.c to LLVM IR and then importing it into MLIR to see
+// what the attributes look like. We could automate this and embed the
+// attributes in the compiler binary but due to the differences in 32/64-bit
+// pointer widths some manual massaging may still be required (or we just embed
+// both).
+//
+// $ clang -emit-llvm -Iruntime/src/ \
+//     runtime/src/iree/hal/local/executable_library_demo.c -g -S \
+//     --target=x86_64-pc-windows-elf
+// $ mlir-translate --import-llvm executable_library_demo.ll
```

Comment:
Sort of dreading keeping this code up to date... but I think the tradeoffs here are acceptable.

Reply:
Ditto - we can always yank it out if it becomes annoying but since this is part of the public ABI I convinced myself we really _shouldn't_ be changing it frequently anyway :)